shiny/ui/_chat.py

```diff
@@ -738,34 +740,57 @@ def _store_message(
 
         return msg
 
-    def _get_trimmed_messages(
-        self,
-        *,
-        token_limits: tuple[int, int] | None = (4096, 1000),
+    @staticmethod
+    def _trim_messages(
+        messages: tuple[StoredMessage, ...],
+        token_limits: tuple[int, int] = (4096, 1000),
     ) -> tuple[StoredMessage, ...]:
-        messages = self._messages()
-
-        if token_limits is None:
-            return messages
-
-        # Can't trim if we don't have token counts
-        token_counts = [m["token_count"] for m in messages]
-        if None in token_counts:
-            return messages
 
-        token_counts = cast("list[int]", token_counts)
+        system_token_count: int = 0
+        token_counts: list[int] = []
+        system_messages: list[str] = []
+        other_messages: list[str] = []
+        for m in messages:
+            count = m["token_count"]
+            # Count can be None if the tokenizer is None
+            if count is None:
+                return messages
+            content = m["content_server"]
+            if m["role"] == "system":
+                system_messages.append(content)
```

Comment:
It looks like `system_messages` and `other_messages` are only used for their `len()`, below, so these can be replaced with simple counters instead of lists.

And if you do that, `content` can be removed entirely.